Sources/SKSwiftPMWorkspace/SwiftPMWorkspace.swift

```diff
@@ -293,6 +323,35 @@ extension SwiftPMWorkspace: SKCore.BuildSystem {
 
     return nil
   }
+
+  /// An event is relevant if it modifies a file that matches one of the file rules used by the SwiftPM workspace.
+  private func isFileEventRelevant(event: FileEvent) -> Bool {
+    // FIXME: fileRules should also include self.workspace.additionalFileRules and toolsVersion should be self.workspace.currentToolsVersion
+    let fileRules = FileRuleDescription.builtinRules
+    let toolsVersion = ToolsVersion.currentToolsVersion
+
+    guard let fileURL = event.uri.fileURL else {
+      return false
+    }
+    let filePath = AbsolutePath(fileURL.path)
+    switch event.type {
+    case .created, .deleted:
+      return fileRules.contains(where: { $0.match(path: filePath, toolsVersion: toolsVersion) })
+    case .changed:
+      return false
+    default: // Unknown file change type
+      return false
+    }
+  }
```

Comment:
@neonichu Could you take a look whether this code makes sense? The intention is to match only files which might impact the build settings computed by SwiftPM. If you think this is correct, I’ll open a PR on SwiftPM to make `Workspace.additionalFileRules` and `Workspace.currentToolsVersion` public.

Reply:
I think it might be better to find a way to refactor `TargetSourcesBuilder` in SwiftPM to provide this kind of information instead of interacting with file rules directly.

Reply:
Do you have any particular path in mind how `Workspace` should expose the `TargetSourcesBuilder`? AFAICT it only creates an ad-hoc `PackageBuilder`, which in turn creates an ad-hoc `TargetSourcesBuilder`.

Reply:
That's a good question, do you have an existing package graph at this point from which you could get a `TargetDescription` to pass down? If yes, maybe just a new function on `Workspace` that gets package identity etc and the target description and returns a configured `TargetSourcesBuilder`? Additionally, I think target sources builder would need a function that would decide whether a singular file path is part of the given package or not.

Reply:
Ah you're right, I wonder whether we could change `TargetSourcesBuilder` to take a `ResolvedTarget` instead. This is turning out to be a bit of a bigger change than I originally thought.

I guess another option could be that we add a new function to `Workspace` to which you pass all the info you have and for now it would just do a simple check using file rules, but we could later refine it to be more accurate. That way we can decouple your PR from making bigger changes to SwiftPM.

BTW, `ManifestJSONParser.parseTarget` is the general purpose target creation, because it is just done as part of manifest loading. We run the compiled manifest, it produces JSON and then we load that JSON.

Reply:
After some offline discussion with @neonichu, we have decided to only watch files matching extensions specified in `FileRuleDescription.builtinRules` in the client. Additionally, I’m adding a function to SwiftPM that will allow more granular filtering based on e.g. the directory structure of a package in the future: https://github.com/apple/swift-package-manager/pull/4016